### PR TITLE
Repair 'helm init --wait': context deadline exceeded

### LINKED ISSUE #3379: helm init --wait gets a context deadline exceeded
```
Running helm init --wait --upgrade --service-account tiller
$HELM_HOME has been configured at /root/.helm.
Error: could not ping Tiller: context deadline exceeded
```
Could be related to using upgrade?

v2.8.0

### PR DESCRIPTION
**Pull request by alexey-igrychev**
Wait now works that way:
* Connect to kubernetes api and wait till tiller pod appeared:
    * Poll every half a second for tiller pod name.
    * Timeout is 5 minutes by default (`--wait-timeout`).
* Create connection over grpc and execute PingTiller.

Also:
* `setupConnection` function is not taking arguments now to be used for `helm init --wait`.
* Hook for cmd renamed to `setupConnectionCobraPreRunHook`.
* PingTiller grpc method is now generated from `tiller.proto`.
* `github.com/golang/protobuf` in glide has been updated to the commit where generated comment `DO NOT EDIT` is in the right place.

Accepted PR https://github.com/kubernetes/helm/pull/3238 problems:
* PingTiller method adds directly into generated tiller.pb.go file (with `DO NOT EDIT` comment). Thus after running `make protoc` PingTiller disappears from `tiller.pb.go` file leading to `make build` errors.
* Also this code is not working because client object doesn't have information about TillerHost so pingTiller uses default host and port and always returns error `Error: could not ping Tiller: context deadline exceeded`.

fixes https://github.com/kubernetes/helm/issues/3379

### CODE REVIEW

File: pkg/helm/client.go [hunk 1]
```diff
@@ -474,14 +474,16 @@ func (h *Client) test(ctx context.Context, req *rls.TestReleaseRequest) (<-chan
...
+func (h *Client) pingTiller(ctx context.Context) error {
 	c, err := h.connect(ctx)
 	if err != nil {
 		return err
 	}
 	defer c.Close()
 
 	rlc := rls.NewReleaseServiceClient(c)
-	return rlc.PingTiller(ctx)
+	_, err = rlc.PingTiller(ctx, &rls.PingTillerRequest{})
```
Comment (bacongobbler) [inline pkg/helm/client.go]: Hey @TheOriginalAlex, would you mind taking a look at this? Was there a way in proto code that we could actually send a request over gRPC without the need to send a request body? That would retain backwards compatibility, but I'm not too concerned given that this is a relatively new feature to Helm.

Thanks!

File: cmd/helm/helm.go [hunk 2]
```diff
@@ -166,7 +166,28 @@ func markDeprecated(cmd *cobra.Command, notice string) *cobra.Command {
 	return cmd
 }
 
-func setupConnection(c *cobra.Command, args []string) error {
+func waitTillerPodAndSetupConnection(waitTimeout int) error {
+	_, client, err := getKubeClient(settings.KubeContext)
```
Comment (michelleN) [inline cmd/helm/helm.go]: I think my only nit is that we're grabbing a kube client here and again during `setupConnection()` but I can understand how this is unavoidable given the way things are. Let me know if you have any ideas for improving that in the future given you're now familiar with the code.

**Review by michelleN — COMMENTED** [review]
I tried to test this manually but I got an Image Pull Err from Kubernetes when trying to install Tiller on the cluster.
```
  24s		8s		2	kubelet, minikube	spec.containers{tiller}	Warning		Failed			Failed to pull image "gcr.io/kubernetes-helm/tiller:v2.8": rpc error: code = Unknown desc = Error response from daemon: manifest for gcr.io/kubernetes-helm/tiller:v2.8 not found
```
I don't understand why it's trying to pull 2.8 vs 2.8.0 which exists. cc/ @bacongobbler

File: cmd/helm/helm.go (hunk 2 above)
Comment (alexey-igrychev) [reply cmd/helm/helm.go]: @michelleN I propose the following solution, https://github.com/kubernetes/helm/pull/3506/commits/c4a01b7229708715a2d680142d744359bc70fc50.

**Comment by bacongobbler** [comment]
@michelleN I've seen that issue happen before and that seems to happen with custom client builds IIRC. You need to set the tiller image manually via `helm init --tiller-image` for it to work on dev builds :(

**Review by bacongobbler — COMMENTED** [review]
the connection handling refactor looks a LOT better. Thanks! I'll just give this a whirl before I give a final LGTM. :)

**Review by bacongobbler — APPROVED** [review]
Works as intended! Thanks!

```
><> helm init --wait --tiller-image quay.io/bacongobbler/tiller:canary
$HELM_HOME has been configured at /home/bacongobbler/.helm.

Tiller (the Helm server-side component) has been installed into your Kubernetes Cluster.
Happy Helming!
```

I observed the tiller pod come up in another window while waiting for Helm to receive the response.

File: pkg/helm/client.go (hunk 1 above)
Comment (TheOriginalAlex) [reply pkg/helm/client.go]: I was just sending a "Ping" string earlier, sending this empty struct does the same job, I think.
https://github.com/kubernetes/helm/blob/master/pkg/proto/hapi/services/tiller.pb.go#L1084

And the server currently just returns "Pong" no matter what the input is, so I don't think this breaks backwards compatibility.
https://github.com/kubernetes/helm/blob/master/pkg/proto/hapi/services/tiller.pb.go#L1314

**Comment by helgi** [comment]
Needs a rebase

**Comment by alexey-igrychev** [comment]
@helgi It's done

**Comment by fibonacci1729** [comment]
As an alternative to generating the PingTiller gRPCs, it might be worth leveraging the health checking support builtin into gRPC. (see: https://github.com/grpc/grpc/blob/master/doc/health-checking.md)

Example Usage: ```
// In k8s.io/helm/cmd/tiller/health.go
package main

import (
    healthpb "google.golang.org/grpc/health/grpc_health_v1"
    "google.golang.org/grpc/health"
)

func WithHealth(setup func() error) {
      s := health.NewServer()
      s.SetServingStatus("Tiller", healthpb.NOT_SERVING)
     if err := setup(); err != nil {
          // handler err....
     }
     s.SetServingStatus("Tiller", healthpb.SERVING)
}
```

**Comment by helgi** [comment]
@fibonacci1729  does that approach impact any of the functionalities in this PR?

**Comment by bacongobbler** [comment]
@helgi from our own testing, no. The end result is the same. It's a much cleaner interface overall. We're hacking on the PR here: https://github.com/bacongobbler/helm/tree/init-wait-fix

**Comment by helgi** [comment]
@bacongobbler much cleaner, yeah. Nice

**Pull request closed without merging**